torch/csrc/jit/codegen/cuda/ops/alias.cpp

```diff
@@ -193,6 +194,73 @@ TensorView* unsqueeze(TensorView* x, int dim) {
   return broadcast(x, broadcast_axes);
 }
 
+TensorView* permute(TensorView* x, const std::vector<int64_t>& new2old) {
+  auto inp_domain = TensorDomain::noReductions(x->getMaybeRFactorDomain());
+  std::vector<IterDomain*> out_domain(inp_domain.size());
+
+  auto normalized_new2old =
+      ir_utils::normalizeNew2Old(new2old, inp_domain.size());
+
+  for (const auto i : c10::irange(out_domain.size())) {
+    auto in_id = inp_domain[new2old[i]];
+    out_domain[i] = in_id->cloneWithoutRFactor();
+  }
+
+  TensorView* out_tensor = IrBuilder::create<TensorView>(
+      IrBuilder::create<TensorDomain>(
+          out_domain, std::vector<bool>(out_domain.size(), true)),
+      x->getDataType().value());
+  IrBuilder::create<TransposeOp>(out_tensor, x, normalized_new2old);
+  return out_tensor;
+}
+
+TensorView* transpose(TensorView* x, int64_t dim0, int64_t dim1) {
+  const auto ndims = static_cast<int>(x->domain()->noReductions().size());
+
+  if (dim0 < 0) {
+    dim0 = ndims + dim0;
+  }
+
+  if (dim1 < 0) {
+    dim1 = ndims + dim1;
+  }
+
+  TORCH_CHECK(
+      dim0 >= 0 && dim0 <= ndims, "Invalid transpose dimension 0: ", dim0);
+
+  TORCH_CHECK(
+      dim1 >= 0 && dim1 <= ndims, "Invalid transpose dimension 1: ", dim1);
+
```

Comment:
nitpick: wasn't the negative dim handling done in `ir_utils::normalizeNew2Old` already, this feels like redundant checks, but does give us nice error message.

Reply:
Dimension canonicalization only wraps around once.
If you had a 3D tensor but specified `dim0 = -4`, then the check will fail.

Reply:
Not saying that we need to change the behavior, both pytorch & numpy only wraps once and errors out for out of range.

`Dimension canonicalization` and the check in `transpose` does exactly the same thing, so it's a redundant check.